components/Home/HomeDownloadButton.tsx

```diff
@@ -1,40 +1,34 @@
 import LocalizedLink from '../LocalizedLink';
 import { useNextraContext } from '../../hooks/useNextraContext';
 import { getNodejsChangelog } from '../../util/getNodeJsChangelog';
-import type { NodeVersionData } from '../../types';
 import type { FC } from 'react';
+import type { NodeRelease } from '../../types';
 
-type HomeDownloadButtonProps = Pick<
-  NodeVersionData,
-  'isLts' | 'node' | 'nodeMajor' | 'nodeNumeric'
->;
-
-const HomeDownloadButton: FC<HomeDownloadButtonProps> = ({
-  node,
-  nodeMajor,
-  nodeNumeric,
+const HomeDownloadButton: FC<NodeRelease> = ({
+  major,
+  version,
+  versionWithPrefix,
   isLts,
 }) => {
   const {
     frontMatter: { labels },
   } = useNextraContext();
 
-  const nodeDownloadLink = `https://nodejs.org/dist/${node}/`;
-  const nodeApiLink = `https://nodejs.org/dist/latest-${nodeMajor}/docs/api/`;
+  const nodeDownloadLink = `https://nodejs.org/dist/${versionWithPrefix}/`;
```

Comment:
@ovflowd In the preview, I saw that this link was not appended with the filename.
In the preview -> https://nodejs.org/dist/v18.16.0/
In nodejs.org -> https://nodejs.org/dist/v18.16.0/node-v18.16.0.pkg

I tried searching but can't find which code is appending the filename.
Do you know about it? 👀

Reply:
Seems like in this function [`detectEnviromentAndSetDownloadOptions`](https://github.com/nodejs/nodejs.org/blob/main/public/static/js/legacyMain.js).

The missing filename is due to race condition.
In nodejs.org, the link was rendered in Server Side. So when that function is invoked, the [query selector](https://github.com/nodejs/nodejs.org/blob/main/public/static/js/legacyMain.js#L121) can detect it and append the filename. But this is not the case since we are fetching the Node releases data in Client Side now.

I am thinking of changing to use [`useDownloadLink`](https://github.com/nodejs/nodejs.org/blob/main/hooks/useDownloadLink.ts) hook. The `legacyMain.js` doesn't seem patch-able to solve this problem. Technically we can put a `sleep` to wait but it is not right I think.

Reply:
@HinataKah0 I guess the best approach here is to get rid of the code from `legacyMain.js` that "replaces the links/append the links" and do that directly on our React Components as we do on `nodejs.dev`.

Afaik we also have a hook on nodejs.dev to detect the environment and we can migrate that.

I completely forgot about this legacy stuff 🤦 would you mind adding this to this PR?

Reply:
@HinataKah0 just a friendly reminder that the link "patching" doesn't only happen on the Home page but also the Download one.

So you might need to also patch something on the Download Components

Reply:
Also, I wonder why the Node Releases are only coming on client-side which the new approach of this PR, shouldn't SWR also run on the server-side?

Reply:
My understanding is that `useSWR` is Client Side by default. If we need SSR, we need to configure it.
https://swr.vercel.app/docs/with-nextjs

Reply:
Let's configure it for SSR so that it can work on SSR too if SSR is supported. But we can do that on a different PR. (I can do that)

Let's for now just ensure all download stuff is working and get rid of the code from legacyMain

Reply:
Initially I wanted to do it naively because it might not worth the effort to refactor since it's only used by legacy website.

But, we can introduce detect bitness functionality to `DownloadCards` component in the new website as well.
Currently it's using `x86` by default, see [this](https://github.com/nodejs/nodejs.dev/blob/main/src/components/DownloadComponents/DownloadCards/index.tsx#L34).

So, let's refactor our detect OS hook & functions a bit.
I am thinking of:
1. Bring back `useDetectOS` hook. This hook ensures that the logic is invoked in Client Side and it provides the fallback as well.
2. Refactor `useDownloadLink` hook to use `useDetectOS` hook.
3. Remove [this](https://github.com/nodejs/nodejs.org/blob/main/util/downloadUrlByOS.ts#L18-L20) and put it in `useDetectOS` hook.

1 & 3 has been done in 656f2bf9c0cd5a068c3ab55b01fafe2056010dfc.

Reply:
@HinataKah0 why do we need the `useDetectOS` hook? Afaik there's nothing the hook does that is reactive.

Reply:
Sorry, I don't get what did you mean by "reactive"?

In legacy website, we need to get both `os` and `bitness` to create home download buttons text (the "Download for ..." text). But, `useDownloadLink` hook only returns the resulting download link.

New website doesn't...
But I believe it's a good functionality to have here, right now the link is not patched (default to `x86`).
![image](https://github.com/nodejs/nodejs.org/assets/128208841/e8380ca4-fa20-4b81-b720-e86f2f446f42)

Why hook?
Because we need to use either `useSWR` or `useEffect` to force the rendering to be done in Client Side.
I chose to keep using `useSWR` because of the built-in cache.

Reply:
> Sorry, I don't get what did you mean by "reactive"?

Hooks usually are done to manage side-effects. In this case, we don't have any side-effect. The whole `detectOS` queries the `window` (both userAgent and bitness API).

It's important to mention that these values will never change during a React life-cycle. Hence there's no need for this to be a hook.

> In legacy website, we need to get both os and bitness to create home download buttons text (the "Download for ..." text). But, useDownloadLink hook only returns the resulting download link.

I'm not questioning that.

> New website doesn't...
But I believe it's a good functionality to have here, right now the link is not patched (default to x86).

We don't need a React Hook for that, that's what I'm saying :)

> Because we need to use either useSWR or useEffect to force the rendering to be done in Client Side.
I chose to keep using useSWR because of the built-in cache.
FYI, using navigation in Server Side will make it crash.

We don't need a Hook. It is true that `useEffect` only runs on the client-side, and can be useful if you want to enforce that something only runs by the client-side.

But that's not even what you're doing [here](https://github.com/nodejs/nodejs.org/pull/5365/files#diff-a5fd39f0144c85ce1e889a4fb3e7d207a01f9717b6c516c05d479a23f1bc4046R24). `useMemo` is called both on the client-side and server-side.

So if you want these values to be **only** evaluated on the client-side, you need to refactor this codebase, and actually store the information within a `react state` (useState).

There's no need to use SWR here at all. As we're not "fetching" data. Just accessing `window` APIs.

Reply:
> I guess it's easier to explain my intent using code 😅
If we choose not to refactor it that way, I can just revert it...

In general yes, but that's not the case here. I understood your code, and I'm still hesitant that this is a good approach.

If the intent was to have a client-side-only evaluation of `window` user agent and bitness, using `useSWR` and `useMemo` is still incorrect.

Reply:
For example you could do:

```tsx
export const useDetectOS = (default: OS = 'OTHER') => {
  const [osInfo, setOSInfo] = useState({ os: default, bitness: 86 });

  useEffect(() => setOSInfo({ .... }), []);

  return osInfo;
}
```

Note: `bitness` should be a number and not a string imho.

Reply:
I tried googling for the word "reactive" + "react" but I wasn't sure if it's what you meant lol
Thanks for explaining it!

> these values will never change during a React life-cycle

Yes, it won't...

Hmm, I think I'll keep it simple with `useState` + `useEffect` in both pages for now.
Since refactoring it is also out of this PR's scope. I don't think it's wise to deep dive into it here. :)

Is it okay with you?

Reply:
I think useDetectOS being the simple useEffect and useState that comes from detectOS and getBitness is more than fine.

I know that this is out of the scope of the PR, but since the changes actually break current production behaviour, we **need** to provide a replacement that is compatible with the changes, meaning:

1. Replace all the usages that depend on legacyMain to use our hooks/apis that we've introduced
2. Remove the code from legacyMain

I know these steps pretty much became an issue because I've suggested for you to instead of add the new stuff on top of the existing stuff to actually replace the existing stuff, but I believe for the long-term this approach would avoid problems in the future.

I appreciate your patience on the subject!!

Reply:
`useDetectOS` is fine with me as well...
But I am not sure if we want to keep the `useDownloadLink`, because I think it can just be a function IMO once we bring back `useDetectOS`... I am more inclined towards the latter.

What I meant by out of this PR's scope was actually about `useDownloadLink` (I phrased it badly).

> I appreciate your patience on the subject!!

Thanks! I am also amazed with you as well :)
I am happy as I am learning along the way

Reply:
If you feel like removing the useDetectDownloadLink that makes more that sense. That hook always struct odd to me 😅